src/cryptonote_core/loki_name_system.cpp

```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
+        return false;
+
+      // Characted preceeding suffix must be alphanumeric
+      char_preceeding_suffix = name_suffix - 1;
+      if (check_condition(!char_is_alphanum(char_preceeding_suffix[0]), reason, "LNS type=", type ,", specifies mapping from name->value where the character preceeding the <char>.loki is not alphanumeric, char=", char_preceeding_suffix[0], ", name=", name))
         return false;
     }
+  }
+  else
+  {
+    // SESSION or WALLET
+    // Name has to start with a (alphanumeric or underscore), and can have (alphanumeric, hyphens or underscores) in between and must end with a (alphanumeric or underscore)
+    // ^[a-z0-9_]([a-z0-9-_]*[a-z0-9_])?$
 
-    // Must start with alphanumeric
-    if (check_condition(!char_is_alphanum(name[0]), reason, "LNS type=lokinet, specifies mapping from name -> value where the name does not start with an alphanumeric character, name=", name))
+    // Must start with (alphanumeric or underscore)
+    if (check_condition(!(char_is_alphanum(name.front()) || name.front() == '_'), reason, "LNS type=", type, ", specifies mapping from name->value where the name does not start with an alphanumeric or underscore character, name=", name))
       return false;
 
-    // Must end with .loki
-    char const SUFFIX[]     = ".loki";
-    char const *name_suffix = name.data() + (name.size() - loki::char_count(SUFFIX));
-    if (check_condition((memcmp(name_suffix, SUFFIX, loki::char_count(SUFFIX)) != 0), reason, "LNS type=lokinet, specifies mapping from name -> value where the name does not end with the domain .loki, name=", name))
+    // Must NOT end with a hyphen '-'
+    if (check_condition(!(char_is_alphanum(name.back()) || name.back() == '_'), reason, "LNS type=", type, ", specifies mapping from name->value where the last character is a hyphen '-' which is disallowed, name=", name))
       return false;
+    char_preceeding_suffix = name.data() + (name.size() - 1);
+  }
 
-    // Characted preceeding suffix must be alphanumeric
-    char const *char_preceeding_suffix = name_suffix - 1;
-    if (check_condition(!char_is_alphanum(char_preceeding_suffix[0]), reason, "LNS type=lokinet, specifies mapping from name -> value where the character preceeding the <char>.loki is not alphanumeric, char=", char_preceeding_suffix[0], ", name=", name))
+  char const *begin = name.data() + 1;
+  char const *end   = char_preceeding_suffix;
+  for (char const *it = begin; it < end; it++) // Inbetween start and preceeding suffix, (alphanumeric, hyphen or underscore) characters permitted
+  {
+    char c = it[0];
+    if (check_condition(!(char_is_alphanum(c) || c == '-' || c == '_'), reason, "LNS type=", type, ", specifies mapping from name->value where the domain name contains more than the permitted alphanumeric or hyphen characters name=", name))
```

Comment:
I think the `c == '_'` here should be `(type != lokinet && c == '_')` to disallow `_`'s from lokinet domains.

Reply:
I've separated the cases for clarity